File: Depth-First Search/Graph/LET_no980_Unique Paths III.py
class Solution:
    def uniquePathsIII(self, grid: List[List[int]]) -> int:
        m, n, empties = len(grid), len(grid[0]), 1
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    x, y = i, j
                if grid[i][j] == 0:
                    empties += 1
        self.res = 0
        self.dfs(grid, x, y, empties)
        return self.res
    
    def dfs(self, grid, x, y, empties):
        if not (0 <= x < len(grid) and 0 <= y < len(grid[0]) and grid[x][y] >= 0 and empties >= 0):
            return
        if grid[x][y] == 2:
            self.res += empties == 0
            return
        
        grid[x][y] = -2
        self.dfs(grid, x - 1, y, empties - 1)
        self.dfs(grid, x, y - 1, empties - 1)
        self.dfs(grid, x + 1, y, empties - 1)
        self.dfs(grid, x, y + 1, empties - 1)
        grid[x][y] = 0

File: Depth-First Search/Graph/LET_no980_Unique Paths III.cpp
class Solution {
private:
    int res, m, n, x, y;
public:
    int uniquePathsIII(vector<vector<int>>& grid) {
        m = grid.size(); n = grid[0].size(); res = 0;
        int empties = 1;
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1) {
                    x = i; y = j;
                }
                if (grid[i][j] == 0)
                    empties += 1;
            }
        }
        
        dfs(grid, x, y, empties);
        return res;
    }
    
    void dfs(vector<vector<int>>& grid, int i, int j, int empties) {
        if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] < 0 || empties < 0)
            return;
        if (grid[i][j] == 2) {
            res += empties == 0;
            return;
        }
        
        grid[i][j] = -2;
        dfs(grid, i - 1, j, empties - 1);
        dfs(grid, i + 1, j, empties - 1);
        dfs(grid, i, j - 1, empties - 1);
        dfs(grid, i, j + 1, empties - 1);
        grid[i][j] = 0;
    }
};